### ISSUE TITLE
Nassau `compute_through_stem` does not schedule optimally

### ISSUE THREAD

**Issue by JoeyBF** [SpectralSequences/sseq#105]
For smaller stems it doesn't matter but for large computations this leads to a lot of wasted time. Here is an example from commit [5da65ea](https://github.com/JoeyBF/sseq/commit/5da65ea845204df9cb2ede4c88e01d34fa01db62).

```
$ cargo run --example resolve_through_stem --release --no-default-features --features nassau,concurrent,logging -- S_2 target/n 100 52
// snip 18756 lines
[     0.315319 s] Computed bidegree (68, 6) with A(2), num new gens = 1, density = 23.27%
[     0.000005 s] [   40.741 MiB/s] Written quasi-inverse for bidegree (85, 2) and signature [4, 3, 0, 1], with Algebra with profile [3, 2, 1, 1]
[     0.000027 s] [  532.081 MiB/s] Written quasi-inverse for bidegree (72, 5) and signature [4, 1, 0], with A(2)
[     0.000039 s] [  232.658 MiB/s] Written quasi-inverse for bidegree (77, 3) and signature [3, 2, 1], with A(2)
[     0.000257 s] [  212.798 MiB/s] Written quasi-inverse for bidegree (67, 7) and signature [0, 0, 0], with Algebra with profile [2, 1, 1]
[     0.000019 s] [  355.995 MiB/s] Written quasi-inverse for bidegree (77, 3) and signature [4, 2, 1], with A(2)
[     0.000049 s] [  370.579 MiB/s] Written quasi-inverse for bidegree (69, 6) and signature [0, 0, 0], with A(2)
[     0.000036 s] [  379.300 MiB/s] Written quasi-inverse for bidegree (72, 5) and signature [5, 1, 0], with A(2)
[     0.000025 s] [  516.419 MiB/s] Written quasi-inverse for bidegree (72, 5) and signature [6, 1, 0], with A(2)
[     0.000009 s] [   56.058 MiB/s] Written quasi-inverse for bidegree (85, 2) and signature [5, 3, 0, 1], with Algebra with profile [3, 2, 1, 1]
[     0.000069 s] [  100.825 MiB/s] Written quasi-inverse for bidegree (77, 3) and signature [5, 2, 1], with A(2)
[     0.000015 s] [  424.611 MiB/s] Written quasi-inverse for bidegree (77, 3) and signature [6, 2, 1], with A(2)
[     0.000008 s] [   15.160 MiB/s] Written quasi-inverse for bidegree (85, 2) and signature [6, 3, 0, 1], with Algebra with profile [3, 2, 1, 1]
[     0.000014 s] [  453.906 MiB/s] Written quasi-inverse for bidegree (77, 3) and signature [7, 2, 1], with A(2)
[     0.000009 s] [   69.957 MiB/s] Written quasi-inverse for bidegree (85, 2) and signature [7, 3, 0, 1], with Algebra with profile [3, 2, 1, 1]
[     0.000028 s] [  390.799 MiB/s] Written quasi-inverse for bidegree (72, 5) and signature [7, 1, 0], with A(2)
[     0.000024 s] [  371.493 MiB/s] Written quasi-inverse for bidegree (77, 3) and signature [0, 3, 1], with A(2)
[     0.000038 s] [  485.103 MiB/s] Written quasi-inverse for bidegree (69, 6) and signature [1, 0, 0], with A(2)
[     0.000055 s] [  292.490 MiB/s] Written quasi-inverse for bidegree (69, 6) and signature [2, 0, 0], with A(2)
[     0.000023 s] [  286.862 MiB/s] Written quasi-inverse for bidegree (77, 3) and signature [1, 3, 1], with A(2)
[     0.000026 s] [  516.302 MiB/s] Written quasi-inverse for bidegree (72, 5) and signature [0, 2, 0], with A(2)
[     0.000036 s] [  450.707 MiB/s] Written quasi-inverse for bidegree (69, 6) and signature [3, 0, 0], with A(2)
[     0.000041 s] [  170.646 MiB/s] Written quasi-inverse for bidegree (77, 3) and signature [2, 3, 1], with A(2)
[     0.000008 s] [   14.510 MiB/s] Written quasi-inverse for bidegree (85, 2) and signature [0, 0, 1, 1], with Algebra with profile [3, 2, 1, 1]
[     0.000005 s] [  137.939 MiB/s] Written quasi-inverse for bidegree (85, 2) and signature [1, 0, 1, 1], with Algebra with profile [3, 2, 1, 1]
[     0.000247 s] [  178.082 MiB/s] Written quasi-inverse for bidegree (67, 7) and signature [1, 0, 0], with Algebra with profile [2, 1, 1]
[     0.000022 s] [  289.083 MiB/s] Written quasi-inverse for bidegree (77, 3) and signature [3, 3, 1], with A(2)
[     0.000033 s] [  379.421 MiB/s] Written quasi-inverse for bidegree (69, 6) and signature [4, 0, 0], with A(2)
[     0.000018 s] [  339.590 MiB/s] Written quasi-inverse for bidegree (77, 3) and signature [4, 3, 1], with A(2)
[     0.000027 s] [  478.194 MiB/s] Written quasi-inverse for bidegree (69, 6) and signature [5, 0, 0], with A(2)
[     0.000072 s] [    4.613 MiB/s] Written quasi-inverse for bidegree (85, 2) and signature [2, 0, 1, 1], with Algebra with profile [3, 2, 1, 1]
[     0.000009 s] [   85.064 MiB/s] Written quasi-inverse for bidegree (85, 2) and signature [3, 0, 1, 1], with Algebra with profile [3, 2, 1, 1]
[     0.000051 s] [  220.804 MiB/s] Written quasi-inverse for bidegree (69, 6) and signature [6, 0, 0], with A(2)
[     0.000022 s] [  233.051 MiB/s] Written quasi-inverse for bidegree (77, 3) and signature [5, 3, 1], with A(2)
[     0.000038 s] [  386.307 MiB/s] Written quasi-inverse for bidegree (72, 5) and signature [1, 2, 0], with A(2)
[     0.000016 s] [  326.748 MiB/s] Written quasi-inverse for bidegree (77, 3) and signature [6, 3, 1], with A(2)
[     0.000023 s] [    9.731 MiB/s] Written quasi-inverse for bidegree (85, 2) and signature [4, 0, 1, 1], with Algebra with profile [3, 2, 1, 1]
[     0.000200 s] [  242.142 MiB/s] Written quasi-inverse for bidegree (67, 7) and signature [2, 0, 0], with Algebra with profile [2, 1, 1]
[     0.000014 s] [  280.000 MiB/s] Written quasi-inverse for bidegree (77, 3) and signature [7, 3, 1], with A(2)
[     0.753501 s] Computed bidegree (77, 3) with A(2), num new gens = 1, density = 15.40%
// snip 774 lines
[     0.186444 s] Computed bidegree (53, 19) with A(1), num new gens = 1, density = 9.12%
[     0.000379 s] [   31.502 MiB/s] Written quasi-inverse for bidegree (69, 6) and signature [7, 0, 0], with A(2)
[     0.000053 s] [   15.570 MiB/s] Written quasi-inverse for bidegree (86, 2) and signature [7, 0, 0, 1], with Algebra with profile [3, 2, 1, 1]
[     0.000342 s] [   44.994 MiB/s] Written quasi-inverse for bidegree (73, 5) and signature [2, 2, 0], with A(2)
[     0.000347 s] [   46.822 MiB/s] Written quasi-inverse for bidegree (69, 6) and signature [0, 1, 0], with A(2)
[     0.000376 s] [   43.860 MiB/s] Written quasi-inverse for bidegree (80, 3) and signature [1, 0, 0], with A(2)
[     0.000107 s] [    8.984 MiB/s] Written quasi-inverse for bidegree (52, 20) and signature [0, 0], with A(1)
[     0.000018 s] [   56.007 MiB/s] Written quasi-inverse for bidegree (52, 20) and signature [1, 0], with A(1)
[     0.000313 s] [   44.082 MiB/s] Written quasi-inverse for bidegree (80, 3) and signature [2, 0, 0], with A(2)
[     0.000032 s] [   23.547 MiB/s] Written quasi-inverse for bidegree (52, 20) and signature [2, 0], with A(1)
[     0.000025 s] [   40.428 MiB/s] Written quasi-inverse for bidegree (86, 2) and signature [0, 1, 0, 1], with Algebra with profile [3, 2, 1, 1]
[     0.000182 s] [   70.483 MiB/s] Written quasi-inverse for bidegree (69, 6) and signature [1, 1, 0], with A(2)
// etc.
```

As you can see, the computation of bidegree (69, 6) starts at the right moment, after both (70,5) and (68,6) are computed. However, it suddenly stops after finishing (77, 3) and only resumes much later after finishing (53, 19)

**Comment by dalcde** [comment]
I'm very confused why this is happening. All subalgebras of the same bidegree are computed in a single loop without giving up control. Do we have something that is secretly taking a lock, and it is taking longer than expected?

**Comment by JoeyBF** [comment]
I don't think so. On the grid it happens sometimes that there is only a single ongoing computation, even though many more are available at that point. Checking the threads with top and pstack I see that only one thread is awake and computing its quasi-inverse (as in #102) and the others are sitting on a condvar somewhere in rayon (AFAIK, I haven't manually checked all other threads, I will next time it happens).

**Comment by JoeyBF** [comment]
I did check and all sleeping threads were just waiting on a rayon condvar. I've been working on this for a while now and the more I think about it the less it makes sense. Potentially days of computation are being wasted right now so I would really like to track that bug down soon

**Comment by JoeyBF** [comment]
There's only one place in the inner loop in `step_resolution_with_subalgebra` where we could possibly give up control and it's when calling `get_partial_matrix`. This means that the only control flow I see is

1. Thread 1 is in the inner loop, computing a bidegree
2. Thread 2 finishes its bidegree, exits `step_resolution`, sends a message through the channel, etc.
3. The next time thread 1 enters `get_partial_matrix`, it gets preempted and goes to sleep.

This makes no sense to me. At least that should be a bug in rayon, and I'm not even sure what mechanism would let a thread put its own job back in the queue and go to sleep. I do observe through top that when a bidegree resumes after such a mysterious hang it usually keeps executing on another thread. I spent hours today looking through rayon internals and I see no end in sight

**Comment by JoeyBF** [comment]
Another update: I'm 100% convinced this is an issue in rayon itself. I'm also suspecting that it's due to some weird incompatibilities between the `spawn` and `join` models. I believe this happens when some thread finishes its bidegree and exits the `spawn` statement while other threads are inside of `get_partial_matrix`. The registry messes up its latches and notifications, but that is still speculation right now. I'll keep working on this tomorrow

**Comment by JoeyBF** [comment]
I've found the issue. It's actually described in [this comment](https://github.com/rayon-rs/rayon/issues/630#issuecomment-459943971) by one of the maintainers of rayon.

In our case, we have two kinds of jobs, bidegree jobs that take several hours, and partial_matrix jobs that take several milliseconds. The problem happens in the following situation:
- Thread 1 starts computing bidegree A.
- Thread 1 executes the `par_iter_mut` in `get_partial_matrix`. This causes threads 1 and 2 to start computing partial_matrix jobs. Let's assume this creates exactly two jobs for simplicity.
- Thread 1 finishes, sees that thread 2 is still executing, and starts looking for other jobs to steal.
- Thread 1 picks up a job for bidegree B that was just pushed to the global deque.
- Thread 2 finishes, but by then thread 1 is stuck in bidegree B, and bidegree A stalls. A priority inversion has occured.

I tried implementing one threadpool for Nassau's algorithm and a separate one for `get_partial_matrix` but that doesn't solve the issue for some reason. The only way I see how to fix this would be to have rayon itself implement some kind of "priority" mechanic, where we could request that a given `join` only picks up jobs that have a high enough priority. Then we give the bidegree jobs low priority and the partial_matrix ones a high one.

**Comment by dalcde** [comment]
In this case are do we actually have much idle cpu that is wadted? Or are we "just" computing things in the wrong order?

**Comment by JoeyBF** [comment]
There's a lot of wasted time. It's hard to tell just looking at the logs how many bidegrees are available, but I would estimate it could compute up to six or seven easily, instead of the two or three at a time it's doing right now (not infrequently even just a single one!)

**Comment by JoeyBF** [comment]
I would expect the speedup to be on the order of 4x. I created a new branch https://github.com/JoeyBF/sseq/tree/bandaid with my threadpool attempt for reference

**Comment by dalcde** [comment]
Interesting. I would have thought the cores are utilized anyway, but I assume the get_partial_matrix concurrency has more overhead

**Comment by JoeyBF** [comment]
At larger stems it spends most of its time outside of `get_partial_matrix`, but that's because concurrency makes it reasonably fast. It really does look like the bulk of the time is spent in `write_qi` when the algebra is smaller than A(3).

Interestingly, there's such a bottleneck that when I made `get_partial_matrix` sequential the time to compute to n = 100 *dropped* from 37s to 32s.

**Comment by JoeyBF** [comment]
I checked if my "fix" was any helpful, and I see some confusing data. On one hand, it is faster, as this hyperfine bench shows
```
$ hyperfine -r 5 --prepare "rm -r test" "./resolve_master S_2 test 140 72" "./resolve_bandaid S_2 test 140 72"
Benchmark 1: ./resolve_master S_2 test 140 72
  Time (mean ± σ):     942.171 s ± 14.949 s    [User: 10172.602 s, System: 366.395 s]
  Range (min … max):   932.343 s … 967.884 s    5 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 2: ./resolve_bandaid S_2 test 140 72
  Time (mean ± σ):     895.731 s ±  7.835 s    [User: 8294.496 s, System: 261.053 s]
  Range (min … max):   885.076 s … 903.508 s    5 runs

Summary
  './resolve_bandaid S_2 test 140 72' ran
    1.05 ± 0.02 times faster than './resolve_master S_2 test 140 72'
```

On the other hand, the problem isn't completely fixed because I still see some stalled bidegrees when I run the command in the OP

**Comment by JoeyBF** [comment]
Fixed in #230 (and more optimally in #276)
